feat: Determine ordering of file groups

suremarc opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #7490 .

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

See details in #7490 - this feature helps DataFusion eliminate sorts when files can be shown to be non-overlapping in terms of min/max statistics.

## What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

* Add a new `FileScanConfig::sort_file_groups` method that distribute files via a bin packing algorithm, ensuring that no two files have overlapping statistics
* Make `FileScanConfig::project` check if file groups are sorted when determining projected output orderings
* Add a new internal `MinMaxStatistics` struct that uses the Arrow Row API to efficiently sort & compare file statistics.

## Are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

Yes - there is a unit test and a sqllogictest. 

## Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

Yes - there is a new optional `statistics` field in `PartitionedFile`, which is part of the proposal in #7490. 

There is also the new `FileScanConfig::sort_file_groups` API

Dandandan commented:
/benchmark

Dandandan commented:
/benchmark

Dandandan commented:
/benchmark

Dandandan commented:
@suremarc sorry for the noise, just trying to run the benchmark command!

Dandandan commented:
/benchmark

Dandandan commented:
Running /benchmark once more to see if previous run was noise

gruuya commented:
Fwiw, I've noticed that it seem to have a minor systematic bias against the PR results in general (even if e.g. there are literally no code changes in it). This manifests like above, by reporting a couple of queries being marginally slower.

I think this will be alleviated once a dedicated runner is used, and/or when more benchmark types are included.

Dandandan commented:
Yeah seems noise, different queries running slower in second run.

alamb reviewed (commented):
Thank you @suremarc  -- this looks very cool


@NGA-TRAN  can you help review this PR (we have a version of this code upstream in InfludDB, as well as a specialized "ProgressiveEval" operator that takes advantage of non-overlapping column ranges

NGA-TRAN commented:
I will review this either today or tomorrow

suremarc commented:
@alamb @NGA-TRAN 

Thanks for the speedy feedback - I guess I'll go ahead and mark this as ready for review since it's being reviewed, haha.  

I'm still planning to add an end-to-end test with `ListingTable` but just haven't figured out how to do it yet

NGA-TRAN commented:
Sorry I could not get to this today. I will try to review it tomorrow/later this week

NGA-TRAN commented:
I am reviewing this now

suremarc commented:
@NGA-TRAN I added a sqllogictest and fixed some bugs in the implementation. CI [seems to be passing now](https://github.com/polygon-io/arrow-datafusion/actions/runs/8370873137)

NGA-TRAN commented:
Thanks @suremarc . I am still in the middle of the review. Sorry it takes more time. So far the code looks really good

NGA-TRAN commented on datafusion/core/src/datasource/physical_plan/file_scan_config.rs:
```diff
@@ -194,6 +203,71 @@ impl FileScanConfig {
+
+    /// Attempts to do a bin-packing on files into file groups, such that any two files
+    /// in a file group are ordered and non-overlapping with respect to their statistics.
+    /// It will produce the smallest number of file groups possible.
```
I understand the purpose of this is to split and group files.  Files in each group is sorted and not overlapped. If there are many groups, each of them must overlap with at least one other.

Is this the first step for following PRs to do something with these sorted file groups? Like files in each group will be read sequentially without going thru  SortPreservingMerge; and then data streams of different groups will go thru SortPreservingMerge?

suremarc replied:
If I understand correctly what you're asking, I think it will already do this -- see for example the sqllogictest I added: https://github.com/apache/arrow-datafusion/blob/e9fad54257a13ab9630a0f038fa2b5a1b26aa006/datafusion/sqllogictest/test_files/parquet.slt#L193-L203

The `FileScanConfig` correctly detects that the files are ordered based on their statistics, and this information is propagated to the physical planner. So fortunately this works already. 

As for next steps, it actually just occurred to me that if there are 1000 files that all overlap, after this PR `ListingTable` will produce 1000 file groups which is really bad. So I think I should add a check that only uses the sorted file groups if they produce less than `target_partitions` file groups. I will add that in this PR. 

The other thing we should maybe do is modify `sort_file_groups` to _increase_ the number of file groups if it's lower than `target_partitions`, but I also thought maybe we should let DataFusion's `EnforceDistribution` physical optimizer take care of this, so it might not be necessary. 

Other than that I'm not sure immediately sure what next steps are, without gathering more data from real-world usage. Though, I would like to add some benchmarks to DataFusion for large plans (with thousands of files), as I think the performance of this code might become more apparent

NGA-TRAN replied:
Thanks. I think another possible next step is to move the implementation of `ProgressiveEvalExec` from InfluxDB IOx here and then  someone can use it and the results of this PR to do https://github.com/apache/arrow-datafusion/issues/6672. 
What do you think @alamb ?

alamb replied:
> As for next steps, it actually just occurred to me that if there are 1000 files that all overlap, after this PR ListingTable will produce 1000 file groups which is really bad. So I think I should add a check that only uses the sorted file groups if they produce less than target_partitions file groups. I will add that in this PR.

I agree


> Thanks. I think another possible next step is to move the implementation of ProgressiveEvalExec from InfluxDB IOx here and then someone can use it and the results of this PR to do #6672.

I think it is a good idea @NGA-TRAN  -- @suremarc and anyone else following along `ProgressiveEvalExec` is an operator we implemented in InfluxDB 3.0 to replace `SortPreservingMerge` in plans where we know the inputs don't overlap in sort-key space (what this PR calculates) and that avoids the merge as well as reads the files more sequentially

> Other than that I'm not sure immediately sure what next steps are, without gathering more data from real-world usage. Though, I would like to add some benchmarks to DataFusion for large plans (with thousands of files), as I think the performance of this code might become more apparent

I agree a real world benchmark would be great. As the "running queries on sorted parquet files" is also quite important to InfluxDB 3.0 we can probably help you with that.

Do you have data in mind? If not, we could potentially make something using https://github.com/apache/arrow-datafusion/blob/main/test-utils/src/data_gen.rs#L251 (that sorts on the first few low cardinality columns, for example)

suremarc replied:
@alamb I didn't have any data in mind, so any headstart we can get sounds great.

Regarding the `SortPreservingMerge`, I want to state again that `sort_file_groups` will produce (if possible) only one file group with all files sequentially ordered, so I think the plan will not have a sort-preserving-merge in the first place. I think maybe I just need to make this clearer in a separate sqllogictest. But crucially this means that #6672 would already be fixed by this PR.

NGA-TRAN commented on datafusion/core/src/datasource/physical_plan/file_scan_config.rs:
```diff
@@ -762,6 +836,171 @@ mod tests {
+                ],
+                sort: vec![col("value").sort(true, false)],
+                expected_result: Ok(vec![vec![0, 1], vec![2]]),
+            },
```
Can we add more tests here?
1. The same input but "2" is now in the middle to ensure we sort and group them correctly
2. All three non-overlapped files
3. All three overlapped files
4. Empty input

alamb replied:
Other important cases are additional negative cases
1. The  files have values but no statistics (the groups shouldn't get reordered)
2. One of the files has no statistics

NGA-TRAN reviewed (approved):
The code looks good. I just ask for a few more unit tests and what is the next step.

alamb commented on datafusion/core/src/datasource/listing/mod.rs:
```diff
@@ -67,6 +67,8 @@ pub struct PartitionedFile {
     pub partition_values: Vec<ScalarValue>,
     /// An optional file range for a more fine-grained parallel execution
     pub range: Option<FileRange>,
+    /// Optional statistics that describe the data in this file
```
I think we should also provide some more guidance here for example

```suggestion
    /// Optional statistics that describe the data in this file if known.
    ///
    /// DataFusion relies on these statistics for planning so if they are incorrect
    /// incorrect answers may result.
```

alamb commented on datafusion/core/src/datasource/physical_plan/file_scan_config.rs:
```diff
@@ -138,12 +138,21 @@ impl FileScanConfig {
             Schema::new(table_fields).with_metadata(self.file_schema.metadata().clone()),
         );
+
+        let full_table_schema = {
```
I wonder if we can avoid creating the entire schema here and instead reuse it from elsewhere -- the information should already be present to plan queries I think. As we try to make planning more efficient avoiding re-creating the same schema over and over again I think will be importatn

suremarc replied:
I managed to get rid of creating this schema, fortunately -- we can use projection (the `Vec<usize>`) to recover the positions of the statistics in the original schema.

alamb commented on datafusion/core/src/datasource/physical_plan/file_scan_config.rs:
```diff
@@ -194,6 +203,71 @@ impl FileScanConfig {
+    /// Attempts to do a bin-packing on files into file groups, such that any two files
+    /// in a file group are ordered and non-overlapping with respect to their statistics.
+    /// It will produce the smallest number of file groups possible.
+    pub fn sort_file_groups(
```
maybe we could name this function something more related to the non overlapping nature

Perhaps like `split_groups_by_statistics` or something?

suremarc replied:
I think that is a good name, I'll go with that

alamb commented on datafusion/core/src/datasource/physical_plan/file_scan_config.rs:
```diff
@@ -194,6 +203,71 @@ impl FileScanConfig {
+            projected_schema,
+            flattened_files.iter().copied(),
+        )
+        .map_err(|e| e.context("construct min/max statistics"))?;
```
Having a bit more context would be helpful I think

```suggestion
        .map_err(|e| e.context("construct min/max statistics for sort_file_groups"))?;
```

alamb commented on datafusion/core/src/datasource/physical_plan/mod.rs:
```diff
@@ -482,6 +508,171 @@ fn get_projected_output_ordering(
+                                .min_value
+                                .get_value()
+                                .cloned()
+                                .unwrap_or(ScalarValue::Null),
```
I don't think we can just use `NULL` when the statistics are not present (as NULL will sort somewhere in the sort order, but in reality the data may be anywhere / possibly overlap) -- Thus I think this code needs to return an error in this case

alamb commented on datafusion/core/src/datasource/physical_plan/file_scan_config.rs:
```diff
@@ -762,6 +836,171 @@ mod tests {
+        }
+
+        struct TestCase {
+            #[allow(unused)]
```
why is `allow(unused)` needed?

suremarc replied:
Definitely not needed, this just got left in from earlier

alamb commented on datafusion/core/src/datasource/physical_plan/mod.rs:
```diff
@@ -448,14 +448,40 @@ impl From<ObjectMeta> for FileMeta {
+                return false;
+            }
+
+            let statistics = match MinMaxStatistics::new_from_files(
```
Since creating `MinMaxStatistics` is likely relatively expensive I wonder if there is some way to avoid it being recreated twice

alamb commented on datafusion/core/src/datasource/physical_plan/mod.rs:
```diff
@@ -482,6 +508,171 @@ fn get_projected_output_ordering(
     all_orderings
 }
 
+// A normalized representation of file min/max statistics that allows for efficient sorting & comparison.
```
I think it would help to explain the requirements on this structure for correctness:
1. That all rows in the sort order have statistics for all values

alamb commented on datafusion/sqllogictest/test_files/parquet.slt:
```diff
@@ -169,6 +171,38 @@ SELECT min(date_col) FROM test_table;
+statement ok
+DROP TABLE test_table;
+
+# Do one more test, but order by numeric columns:
```
I would like to request moving the tests for this optimization to its own .slt file (e.g. `sorted_parquet.slt` or something). It is important enough and tricky enough I think to warrant specialized tests

alamb commented on datafusion/sqllogictest/test_files/parquet.slt:
```diff
@@ -169,6 +171,38 @@ SELECT min(date_col) FROM test_table;
+LOCATION 'test_files/scratch/parquet/test_table';
+
+# Check output plan again, expect an "output_ordering" clause in the physical_plan -> ParquetExec:
+# After https://github.com/apache/arrow-datafusion/pull/9593 this should not require a sort.
```
I don't fully understand this comment -- this is a new test added as part of that PR, right?

suremarc replied:
Yeah I guess the wording is confusing since it implies the test existed before this PR. Either way I'm going to make a new sqllogictest so I'll rewrite this comment

alamb reviewed (commented):
Thank you @suremarc  -- this is very cool and definitely a great feature to add. 

Thank you also @NGA-TRAN  for the review

From my perspective, the only thing this PR really needs is some additional tests (especially around missing statistics). Since it is effectively "opt in" (the code won't be activated unless a user explicitly supplies `PartitionedFile::Statistics`) I think it would be ok to merge in after that and then we can file follow on tickets for things like benchmarks, ProgressiveEval, etc

What do you think?

suremarc commented:
@alamb I won't be able to get to this until tomorrow, just letting you know

alamb commented:
> @alamb I won't be able to get to this until tomorrow, just letting you know

No worries - thanks for doing it @suremarc  -- I am traveling this week so I won't have time to review until later this week. 

Thank you so much

alamb commented:
I think this PR is quite nice and it would be great if we could get the tests written and the code polished up. 

Marking as draft as I think this PR is no longer waiting on feedback. Please mark it as ready for review when it is ready for another look

suremarc reviewed (commented):
@alamb @NGA-TRAN 

Sorry for the long delay. I've been busy the last month and there ended up being some bugs with the implementation, particularly around descending sort & projecting physical expressions. 

I've added a sqllogictest with descending sort, projections, sorting on partition columns, and overlapping files. I think it covers the edge cases pretty well. Please feel free to take a look when you have time

alamb commented:
@NGA-TRAN  do you have time to review this PR as well?

alamb commented:
No worries @suremarc  -- I am very excited about this PR. I plan to review it sometime this week (hopefully later today)

NGA-TRAN commented:
I will review this either today or tomorrow

NGA-TRAN commented on datafusion/sqllogictest/test_files/parquet_sorted_statistics.slt:
```diff
@@ -0,0 +1,352 @@
+
+# File 1:
+query ITID
+COPY (SELECT * FROM src_table LIMIT 3)
```
I am a bit surprised this query always return the first 3 rows and its dat tis sorted. Maybe you want to make it deterministic in case this behavior no longer holds in the future by using `SELECT * FROM src_table where int_col <=3 order by int_col`

I think since you make your columns have corresponding increasing data, the order by can be on any column and your data is always sorted on any column

suremarc replied:
I'm not sure what happened here but I think you were looking at an old copy of the code -- GH says this is outdated. The latest version has a sort in this command

NGA-TRAN replied:
Right. See my comment here https://github.com/apache/datafusion/pull/9593#pullrequestreview-2032171364

suremarc replied:
Ah, ok that makes sense

NGA-TRAN commented on datafusion/sqllogictest/test_files/parquet_sorted_statistics.slt:
```diff
@@ -0,0 +1,352 @@
+
+# File 2:
+query ITID
+COPY (SELECT * FROM src_table WHERE int_col > 3 LIMIT 3)
```
Similar as above,  `SELECT * FROM src_table where int_col >=4 and int_col <=6 order by int_col`

NGA-TRAN commented on datafusion/sqllogictest/test_files/parquet_sorted_statistics.slt:
```diff
@@ -0,0 +1,352 @@
+
+# Add another file to the directory underlying test_table
+query ITID
+COPY (SELECT * FROM src_table WHERE int_col > 6 LIMIT 3)
```
Same as above, you want filter that provide deterministic result and data is sorted

NGA-TRAN commented on datafusion/sqllogictest/test_files/parquet.slt:
```diff
@@ -132,8 +132,8 @@ STORED AS PARQUET;
-# Check output plan again, expect an "output_ordering" clause in the physical_plan -> ParquetExec:
-# After https://github.com/apache/arrow-datafusion/pull/9593 this should not require a sort.
+# Check output plan again, expect no "output_ordering" clause in the physical_plan -> ParquetExec,
+# due to there being more files than partitions:
```
Nice negative test

NGA-TRAN reviewed (approved):
Thanks @suremarc for adding tests and all refactoring works. The PR looks good to me

suremarc commented:
I'm not sure why changing a comment caused the tests to start failing.... oof.

alamb commented:
Added API change label as I it adds a new field to `PartitionedFile`

alamb commented on datafusion/core/src/datasource/physical_plan/file_scan_config.rs:
```diff
@@ -194,6 +196,77 @@ impl FileScanConfig {
+            e.context("construct min/max statistics for split_groups_by_statistics")
+        })?;
+
+        let indices_sorted_by_min = {
```
I wonder if we could move this into `statistics itself`  somehing like

```rust
        let indices_sorted_by_min = statistics.indices_sorted_by_min()
```

alamb commented on datafusion/core/src/datasource/physical_plan/file_scan_config.rs:
```diff
@@ -770,6 +843,277 @@ mod tests {
+                sort: vec![col("value").sort(true, false)],
+                expected_result: Ok(vec![vec!["0"], vec!["1"], vec!["2"]]),
+            },
+            TestCase {
```
can we please add a test for a single input file too?

alamb commented on datafusion/core/src/datasource/physical_plan/file_scan_config.rs:
```diff
@@ -770,6 +843,277 @@ mod tests {
+                    false,
+                )]),
+                files: vec![
+                    File::new("0", "2023-01-01", vec![Some((0.00, 0.49))]),
```
I think all these tests also always have the first file with the minimum stastistics value -- can you possibly also test what happens when it is not (aka add a test that runs this test with file ids 2, 1, 0)?

alamb reviewed (commented):
Thank you so much @suremarc  -- this PR is pretty epic

I am a little worried about the lack of coverage in `MinMaxStatistics`, especially around some of the tricky edge cases with projected schemas and projections.  However this PR is already quite large and been open for quite a while

Here is how I suggest we proceed:
1.  Add a config value for this feature and default it to `false` (not enabled by default)
2. Merge this PR
3. File additional tickets / tests before enabling it
4. Add additional test coverage as follow on PRs and do some more testing.

Once we feel good about that we can make a PR to turn it on by default.

Thank you very much @NGA-TRAN  for your reviews

suremarc commented on datafusion/core/src/datasource/physical_plan/mod.rs:
```diff
@@ -473,15 +468,281 @@ fn get_projected_output_ordering(
+    fn new_from_files<'a>(
+        projected_sort_order: &[PhysicalSortExpr], // Sort order with respect to projected schema
+        projected_schema: &SchemaRef,              // Projected schema
+        projection: Option<&[usize]>, // Indices of projection in full table schema (None = all columns)
```
I am not sure what you mean, but `projection` is what was used to produce `projected_schema`. It tells us what position the columns of `projected_schema` would be in the full schema. Does that make it more clear?

alamb replied:
I guess I was thinking of subtle bugs related to when:
1. The schema of the files is different but compatible (e.g. one file as (`time`, `date`, `symbol`) but the other file had (`date`, `symbol`, `time`) for example
2. The query orders by a subset of the columns (e.g. `ORDER BY time`)
3. The query orders by a subset of the columns that is not the sort order (`ORDER BY date`)

suremarc replied:
Oh... I didn't even think about option 1. But I was assuming that the layout of the file statistics should match the table schema and not the individual file's schema. It seems that that's what DataFusion does currently.

suremarc commented:
@alamb I added a config value, and I moved `MinMaxStatistics` to its own module as requested. I wasn't sure if I should delay addressing your feedback on tests to the next PR, since it seems like the suggested plan is to merge this PR first.

alamb commented:
> @alamb I added a config value, and I moved MinMaxStatistics to its own module as requested. I wasn't sure if I should delay addressing your feedback on tests to the next PR, since it seems like the suggested plan is to merge this PR first.

Sorry -- sounds good. I am going to give this PR another look and file some follow on tickets.

alamb reviewed (approved):
Thank you @suremarc  -- epic work. Let's merge this one in and keep iterating on main

Thanks again for sticking with this -- this is very exciting

alamb commented:
Filed https://github.com/apache/datafusion/issues/10336 to track enable this flag by default
